pkg/twcc/twcc.go

```diff
@@ -33,68 +37,118 @@ type Recorder struct {
 // feedback packets.
 func NewRecorder(senderSSRC uint32) *Recorder {
 	return &Recorder{
-		receivedPackets: []pktInfo{},
-		senderSSRC:      senderSSRC,
+		senderSSRC: senderSSRC,
 	}
 }
 
 // Record marks a packet with mediaSSRC and a transport wide sequence number sequenceNumber as received at arrivalTime.
 func (r *Recorder) Record(mediaSSRC uint32, sequenceNumber uint16, arrivalTime int64) {
 	r.mediaSSRC = mediaSSRC
-	if sequenceNumber < 0x0fff && (r.lastSequenceNumber&0xffff) > 0xf000 {
-		r.cycles += 1 << 16
+
+	// "Unwrap" the sequence number to get a monotonically increasing sequence number that
+	// won't wrap around after math.MaxUint16.
+	unwrappedSN := r.sequenceUnwrapper.Unwrap(sequenceNumber)
+	r.maybeCullOldPackets(unwrappedSN, arrivalTime)
+	if r.startSequenceNumber == nil || unwrappedSN < *r.startSequenceNumber {
+		r.startSequenceNumber = &unwrappedSN
 	}
-	r.receivedPackets = insertSorted(r.receivedPackets, pktInfo{
-		sequenceNumber: r.cycles | uint32(sequenceNumber),
-		arrivalTime:    arrivalTime,
-	})
-	r.lastSequenceNumber = sequenceNumber
-}
 
-func insertSorted(list []pktInfo, element pktInfo) []pktInfo {
-	if len(list) == 0 {
-		return append(list, element)
+	// We are only interested in the first time a packet is received.
+	if r.arrivalTimeMap.HasReceived(unwrappedSN) {
+		return
 	}
-	for i := len(list) - 1; i >= 0; i-- {
-		if list[i].sequenceNumber < element.sequenceNumber {
-			list = append(list, pktInfo{})
-			copy(list[i+2:], list[i+1:])
-			list[i+1] = element
-			return list
-		}
-		if list[i].sequenceNumber == element.sequenceNumber {
-			list[i] = element
-			return list
-		}
+
+	r.arrivalTimeMap.AddPacket(unwrappedSN, arrivalTime)
+
+	// Limit the range of sequence numbers to send feedback for.
+	if *r.startSequenceNumber < r.arrivalTimeMap.BeginSequenceNumber() {
+		sn := r.arrivalTimeMap.BeginSequenceNumber()
+		r.startSequenceNumber = &sn
+	}
+}
+
+func (r *Recorder) maybeCullOldPackets(sequenceNumber int64, arrivalTime int64) {
+	if r.startSequenceNumber != nil && *r.startSequenceNumber >= r.arrivalTimeMap.EndSequenceNumber() && arrivalTime >= packetWindowMicroseconds {
+		r.arrivalTimeMap.RemoveOldPackets(sequenceNumber, arrivalTime-packetWindowMicroseconds)
```

Comment:
maybe, this can call `RemoveOldPackets` unconditionally and that API can do the needful. As this is called on every packet, that method can do fast path check as efficiently as possible and return early.

Reply:
I haven't benchmarked this, but I suspect the difference is very minor. In contrast, keeping a direct port of libwebrtc makes it easier to audit that they have the same behavior, so I think even if there is a slight performance hit, it's worth it! It's worth noting that this implementation is O(1) insert (amortized) and O(N) build feedback, which should be much faster than what's on master currently!